### PR TITLE
v.30: pass config file to meilisearch

### LINKED ISSUE #1962: v.30: Pass config file to Meilisearch
Users can now send Meilisearch a configuration file instead of command line options at launch

- Update the "[Customization](https://docs.meilisearch.com/learn/getting_started/quick_start.html#customization)" section of the quick start
- Add new section to the [Configuration](https://docs.meilisearch.com/learn/configuration/instance_options.html) guide

### References
- Issue on core: https://github.com/meilisearch/meilisearch/issues/2558, https://github.com/meilisearch/meilisearch/issues/2800, https://github.com/meilisearch/meilisearch/issues/2801
- Spec: https://github.com/meilisearch/specifications/pull/185
- SME: @curquiza

### PR DESCRIPTION
**Pull request by maryamsulemani97**
closes #1962

### CODE REVIEW

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:140]: This option cannot be used in the config file. Maybe could we mention it here again?

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:15]: I would add some context, like "using environment variables and config file at the same time would lead to..."

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:21]: I would write Meilisearch will read the `./config.toml` file by default, I mean the file in the working directory named `config.toml`

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:36]: I would put it at the end of the section, because if no `./config.toml` file are found OR if you haven't provided any file with `--config-file-path`, `none` is displayed.
This sentence only helps users for debugging but is not necessary to understand the feature.

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:48]: Hum, not sure this is useful to talk about the default file we provide. The users do not need to provide a file like this exactly to make it work. It's really confusing to me to explain this

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:51]: Do we really need to define errors in the config file feature? Is it useful for the users? Looks to much like a specification, not a guide.

Comment (curquiza) [reply learn/configuration/instance_options.md:51]: My opinion is the full section can be removed, let me know what you think @maryamsulemani97 and @guimachiavelli

**Review by curquiza — CHANGES_REQUESTED** [review]

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:51]: I added it here because I though it might be useful, but we can get rid of it

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:48]: It might be helpful for first-time users who just downloaded the file. Maybe we could add this part:
>Some keys are commented out as they are not required when launching Meilisearch, for example, `import_snapshot`, or because the value provided in the file is not the default value like `max_indexing_memory`.

to the config file, if you think it's useful

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:140]: Yes, good idea!

File: learn/configuration/instance_options.md
Comment (curquiza) [reply learn/configuration/instance_options.md:51]: From my user POV I would say it's not useful for me to know all the behavior of the features in case of errors. Errors should be explicit enough to guide me, but the docs don't need to describe them. It makes the docs heavier.
But I would rather have the opinion of gui, maybe he does not agree with me.

File: learn/configuration/instance_options.md
Comment (curquiza) [reply learn/configuration/instance_options.md:48]: The problem is we don't have the context here, we don't know we are talking about the default config file here.

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:21]: "Some keys are commented out as they are not required when launching Meilisearch" => No key are required when launching Meilisearch.

From my POV this is not really useful to "justify" our default config file, but only my opinion

File: .code-samples.meilisearch.yaml
Comment (guimachiavelli) [inline .code-samples.meilisearch.yaml:1042]: Unless I'm mistaken, instance options cannot be configured with an SDK, so perhaps it's better to leave it outside the code sample component?

@brunoocasali, what do you think?

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:13]: I find it a bit weird to say that configuration files are alternatives to cli/env vars, but have it as the first configuration method. I suggest we move it below environment variables.

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:15]: I recommend dropping the "can"s so statements are simple and leave no room for interpretation: env vars overwrite config files. You might also be able to get away with dropping the first clause, but it's your call.

```suggestion
**When used simultaneously, environment variables overwrite the configuration file, and command-line options overwrite environment variables.**
```

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [reply learn/configuration/instance_options.md:21]: I believe "key" here stands for "option", no? In which case, I recommend saying option to avoid misunderstandings.

In any case, I tend to agree with with @curquiza: not sure this paragraph is necessary since we're only describing the file.

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:23]: I think we can drop the article. Also, we don't need to use both `./` and working directory, in my opinion because the working directory is `./`.

```suggestion
By default, Meilisearch will use the `config.toml` in the working directory as the default configuration file. You will see it when you launch Meilisearch:
```

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:42]: Last bit is unnecessary and doesn't connect with the snake_case convention, imo. `import_dumps` throws an error not because it's not following the convention, but because that option doesn't exist.

```suggestion
You can configure all environment variables and CLI options using a dedicated key in the configuration file. These keys follow the [snake case](https://en.wikipedia.org/wiki/Snake_case) convention. For example, `--import-dump` must be named `import_dump` in the file.
```

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:45]: The first sentence lacks context: the only exception to what? the snake_case bit and the warning box make create too much distance between this statement and  "You can configure all environment variables".

```suggestion
Specifying the `config_file_path` key in the configuration file will throw an [error](/learn/configuration/instance_options.md#defining-a-configuration-file-with-a-syntax-error).
```

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [reply learn/configuration/instance_options.md:49]: I agree with @curquiza, I also think this section makes the article too long. All the errors seem to work pretty much like other Meilisearch errors, so there's nothing unexpected that justifies talking about them here, I think.

File: learn/getting_started/quick_start.md
Comment (guimachiavelli) [inline learn/getting_started/quick_start.md:282]: The config file only accepts instance options, right? If so, I think it's better to have it as subheading of `### Instance options` instead of a self-sufficient section like `### Index settings`.

```suggestion
#### Configuration file
```

**Review by guimachiavelli — CHANGES_REQUESTED** [review]

File: learn/getting_started/quick_start.md
Comment (maryamsulemani97) [reply learn/getting_started/quick_start.md:282]: It is a subheading of `Instance options` 👀

Comment (guimachiavelli) [reply learn/getting_started/quick_start.md:282]: Oh, but aren't both of them on the same heading level?

`### Instance options` on L#276

`### Configuration file` on L#282

Both are level three (`###`), but for configuration file to be a subheading it would have to be level four (`####`), no?

Comment (maryamsulemani97) [reply learn/getting_started/quick_start.md]: I get it now! Yes, you're right 🙏

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:72]:
```suggestion
You can override the default location of the configuration file using the `MEILI_CONFIG_FILE_PATH` environment variable or the `--config-file-path` CLI option:
```

**Review by guimachiavelli — CHANGES_REQUESTED** [review]

File: learn/configuration/instance_options.md
Comment (curquiza) [inline learn/configuration/instance_options.md:92]: Just by curiosity? Why do we now explain how to set an environment variable with Windows since we don't do it [here](https://docs.meilisearch.com/learn/configuration/instance_options.html#environment-variables)?

Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:92]: Good catch. I'll update it here too

File: .code-samples.meilisearch.yaml
Comment (brunoocasali) [reply .code-samples.meilisearch.yaml:1042]: Yes, that's correct! The SDK uses an already-initialized instance :)

File: learn/getting_started/quick_start.md
Comment (guimachiavelli) [inline learn/getting_started/quick_start.md:284]: Trying to avoid repeating the same verb twice in quick succession.

```suggestion
Instead of providing command-line options or environment variables, you can use a `.toml` [configuration file](/learn/configuration/instance_options.md#configuration-file). Once downloaded, move it to your working directory and it will be loaded at launch. Like instance options, it affects your entire instance.
```

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:36]: Would it be more accurate/inclusive to say UNIX instead of Linux? What do you think, @curquiza?

File: learn/configuration/instance_options.md
Comment (guimachiavelli) [inline learn/configuration/instance_options.md:122]: What do you think about removing the "using a dedicated option" bit?

```suggestion
You can configure all environment variables and CLI options using the configuration file. These options must be written in [snake case](https://en.wikipedia.org/wiki/Snake_case). For example, `--import-dump` must be named `import_dump` in the file.
```

**Review by guimachiavelli — CHANGES_REQUESTED** [review]
Ok, sorry, I think this is the last review.

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:36]: Note to self: Update [this](https://docs.meilisearch.com/learn/security/master_api_keys.html#protecting-a-meilisearch-instance) if we go with UNIX

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:122]: We're talking about the file format here 🤔  So wouldn't it be better to talk about "options"? We don't mention them anywhere else.

File: learn/configuration/instance_options.md
Comment (curquiza) [reply learn/configuration/instance_options.md:36]: Yes, in this case `Unix` is better

**Review by curquiza — APPROVED** [review]

File: learn/advanced/updating.md
Comment (curquiza) [inline learn/advanced/updating.md:144]: Unix, includes Macos

Comment (curquiza) [inline learn/advanced/updating.md:144]:
```suggestion
::: tab UNIX
```

**Review by curquiza — CHANGES_REQUESTED** [review]

File: learn/advanced/updating.md
Comment (maryamsulemani97) [reply learn/advanced/updating.md:144]: Thanks @curquiza! 🙏

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:75]: What if you launch Meilisearch w/o a configuration file, does the `Config file path` field still appear in the startup message? If so, what is its value?

```suggestion
By default, Meilisearch will look for a `config.toml` file in the working directory. If it is present, it will be used as the configuration file. You can verify this when you launch Meilisearch:
```

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:101]:
```suggestion
::: tab Env
```

File: learn/configuration/instance_options.md
Comment (dichotommy) [reply learn/configuration/instance_options.md:122]: I agree w/ Gui's suggestion, and would go a little further.

I also think that after this paragraph we should provide an example of a full line from a config file. Of course, anyone curious enough can look at the default config file we provide and figure it out from that... but people are lazy! This should make the formatting completely clear.

````suggestion
You can configure any environment variable or CLI option using a configuration file. In configuration files, options must be written in [snake case](https://en.wikipedia.org/wiki/Snake_case). For example, `--import-dump` would be written as `import_dump`.

```toml
import_dump = "./example.dump"
```
````

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:125]:
```suggestion
Specifying the `config_file_path` option within the configuration file will throw an error. This is the only configuration option that cannot be set within a configuration file.
```

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:120]: "file format" = `jpg`, `toml`, etc.
"file formatting" = rules for arranging text
TMYK 😉 

```suggestion
### Configuration file formatting
```

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:140]:
```suggestion
Specifying this option in the configuration file itself will throw an error (assuming Meilisearch is able to find your configuration file).
```

File: learn/getting_started/quick_start.md
Comment (dichotommy) [inline learn/getting_started/quick_start.md:284]: It's not necessary to download a premade configuration file, since users can also easily construct their own based on the formatting rules we provide.

```suggestion
As an alternative to providing command-line options or environment variables at launch, you can configure your Meilisearch instance using a [configuration file](/learn/configuration/instance_options.md#configuration-file) in `.toml` format. If a configuration file is present at the root of your working directory, it will be loaded at launch.
```

**Review by curquiza — APPROVED** [review]

**Comment by curquiza** [comment]
Nice guide @maryamsulemani97 👏

File: learn/configuration/instance_options.md
Comment (maryamsulemani97) [reply learn/configuration/instance_options.md:101]: We use "Environment variable" [here](https://docs.meilisearch.com/learn/security/master_api_keys.html#protecting-a-meilisearch-instance)

Comment (dichotommy) [reply learn/configuration/instance_options.md:101]: I see. Please ignore my suggestion then—it's just some needless optimization 😅

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:89]:
```suggestion

If the `Config file path` is anything other than `"none"`, it means that a config file was successfully located and used to start Meilisearch.

```

File: learn/configuration/instance_options.md
Comment (dichotommy) [inline learn/configuration/instance_options.md:134]:
```suggestion
### Configuration file path
```

**Review by dichotommy — APPROVED** [review]
Added two small comments, let me know if you need another approval after committing 🏂🏻 ⛰️

**Comment by maryamsulemani97** [comment]
bors merge

**Pull request merged**
